### PR TITLE
[Docs][LTO] Updated HowToSubmitABug.rst for LTO crashes

### PR DESCRIPTION
**Pull request by xgupta**

### CODE REVIEW

File: llvm/docs/HowToSubmitABug.rst [hunk 1]
```diff
@@ -153,6 +153,57 @@ Please run this, then file a bug with the instructions and reduced .bc file
...
+
+If you find a bug that crashes llvm in LTO phase (by using -flto option),
+compile your source file to a .bc file by passing
+"``-flto -fuse-ld=lld -Wl,-plugin-opt=save-temps``"
+to clang (in addition to the options you already pass). If you are building
+a project, pass the appropriate CFLAGS, CXXFLAG​S and LDFLAGS for example -
+
+.. code-block:: bash
+
+   export CFLAGS="-flto -fuse-ld=lld" CXXFLAGS="-flto -fuse-ld=lld" LDFLAGS="-Wl,-plugin-opt=save-temps"
```
Comment (mizvekov) [inline llvm/docs/HowToSubmitABug.rst:169]: Would be nice if this covered the clang-cl / lld-link case as well.

**Review by w2yehia — APPROVED** [review]

File: llvm/docs/HowToSubmitABug.rst [hunk 2]
```diff
@@ -153,6 +153,67 @@ Please run this, then file a bug with the instructions and reduced .bc file
...
+
+These flags enable LTO and save temporary files generated during compilation
+for later analysis.
+
+On Windows, you should use lld-link as the linker. Adjust your compilation 
+flags as follows:
+
+.. code-block:: bash
+
+   export CFLAGS="-flto -fuse-ld=lld-link" CXXFLAGS="-flto -fuse-ld=lld-link" LDFLAGS="-Wl,-plugin-opt=save-temps"
```
Comment (mizvekov) [inline llvm/docs/HowToSubmitABug.rst:177]: Does this actually work? The option is spelled `/lldsavetemps` in the COFF linker.

Comment (xgupta) [reply llvm/docs/HowToSubmitABug.rst:177]: I am not sure, I have never tried these on Windows (I mainly work on Linux), can you share the exact command, I think export is not a Windows command also.

Comment (mizvekov) [inline llvm/docs/HowToSubmitABug.rst:177]:
```suggestion
On Windows, you should be using lld-link as the linker. Adjust your compilation 
flags as follows:
* Add `/lldsavetemps` to the linker flags.
* When linking from the compiler driver, Add `/link /lldsavetemps` in order to forward that flag to the linker.

File: llvm/docs/HowToSubmitABug.rst [hunk 3]
```diff
@@ -153,6 +153,67 @@ Please run this, then file a bug with the instructions and reduced .bc file
...
+
+#. a.out.0.0.preopt.bc (Before any link-time optimizations (LTO) are applied)
+#. a.out.0.2.internalize.bc (After initial optimizations are applied)
+#. a.out.0.4.opt.bc (After an extensive set of optimizations)
+#. a.out.0.5.precodegen.bc (After LTO but before translating into machine code)
+
+Execute one of the following commands to identify the source of the problem:
+
+#. ``opt "-passes=lto<O3>" a.out.0.0.preopt.bc``
+#. ``opt "-passes=lto<O3>" a.out.0.2.internalize.bc``
```
Comment (MaskRay) [inline llvm/docs/HowToSubmitABug.rst:189]: For regular LTO, we can keep just `opt "-passes=lto<O3>" a.out.0.2.internalize.bc` and remove the other 2 runs.

File: llvm/docs/HowToSubmitABug.rst [hunk 4]
```diff
@@ -153,6 +153,67 @@ Please run this, then file a bug with the instructions and reduced .bc file
...
+
+Example of reduce.sh Script
+
+.. code-block:: bash
+
+   $ cat reduce.sh
+   #!/usr/bin/env bash
+
+   $HOME/llvm/llvm-project/build/bin/opt "-passes=lto<O3>" $1 -o temp.bc  2>&1 | tee err.log
+   grep "It->second == &Insn" err.log
```
Comment (MaskRay) [inline llvm/docs/HowToSubmitABug.rst:209]: `grep -q ...` and remove the last `exit $?`

File: llvm/docs/HowToSubmitABug.rst (hunk 3 above)
Comment (MaskRay) [reply llvm/docs/HowToSubmitABug.rst:189]: For ThinLTO, it is `*.3.import.bc`

**Comment by xgupta** [comment]
Gentle ping for review!

File: llvm/docs/HowToSubmitABug.rst [hunk 5]
```diff
@@ -153,6 +153,62 @@ Please run this, then file a bug with the instructions and reduced .bc file
 that bugpoint emits.  If something goes wrong with bugpoint, please submit
 the "foo.bc" file and the option that llc crashes with.
 
+LTO bugs
+---------------------------
+
+If you encounter a bug that leads to crashes in the LLVM LTO phase when using
+the `-flto` option, follow these steps to diagnose and report the issue:
```
Comment (MaskRay) [inline llvm/docs/HowToSubmitABug.rst:160]: double backquotes

File: llvm/docs/HowToSubmitABug.rst [hunk 6]
```diff
@@ -153,6 +153,62 @@ Please run this, then file a bug with the instructions and reduced .bc file
...
+
+Compile your source file to a .bc (Bitcode) file with the following flags,
```
Comment (MaskRay) [inline llvm/docs/HowToSubmitABug.rst:162]: quote `.bc`

flags => options.

`-fuse-ld=` is an option, not a flag (boolean option).

File: llvm/docs/HowToSubmitABug.rst [hunk 7]
```diff
@@ -153,6 +153,62 @@ Please run this, then file a bug with the instructions and reduced .bc file
...
+in addition to your existing compilation options:
+
+.. code-block:: bash
+
+   export CFLAGS="-flto -fuse-ld=lld" CXXFLAGS="-flto -fuse-ld=lld" LDFLAGS="-Wl,-plugin-opt=save-temps"
+
+These flags enable LTO and save temporary files generated during compilation
```
Comment (MaskRay) [inline llvm/docs/HowToSubmitABug.rst:169]: options

File: llvm/docs/HowToSubmitABug.rst [hunk 8]
```diff
@@ -153,6 +153,62 @@ Please run this, then file a bug with the instructions and reduced .bc file
...
+for later analysis.
+
+On Windows, you should be using lld-link as the linker. Adjust your compilation 
+flags as follows:
+* Add `/lldsavetemps` to the linker flags.
```
Comment (MaskRay) [inline llvm/docs/HowToSubmitABug.rst:174]: double backquotes

File: llvm/docs/HowToSubmitABug.rst [hunk 9]
```diff
@@ -153,6 +153,62 @@ Please run this, then file a bug with the instructions and reduced .bc file
...
+
+   llvm-reduce --test reduce.sh a.out.0.2.internalize.bc
+
+Example of reduce.sh Script
+
+.. code-block:: bash
+
+   $ cat reduce.sh
+   #!/usr/bin/env bash
+
```
Comment (MaskRay) [inline llvm/docs/HowToSubmitABug.rst:203]:
```
#!/bin/bash -e
path/to/not --crash path/to/opt  ... 2> err.log
grep ... err.log
```

**Review by MaskRay — APPROVED** [review]

**Comment by xgupta** [comment]
Thank you all for the review.

**Pull request merged**